### PR TITLE
feat: configurable summarization LLM, retrieval params, and chunk threshold

### LINKED ISSUE #21: Configurable pipeline: separate summarization LLM and externalized retrieval parameters
## Context
Part of alkem-io/alkemio#1815 — RAG Pipeline Improvements (Phase 2: Chunk & Embedding Tuning)

## Problem
The pipeline has several values that should be tunable in production without code changes, and one significant cost optimization opportunity:

1. **Summarization uses the same LLM as user-facing responses.** `DocumentSummaryStep` and `BodyOfKnowledgeSummaryStep` receive the same `LLMPort` that powers expert/guidance answers. Summarization is a simpler task — it doesn't need the best reasoning model. The original TypeScript `ingest-space` already used a separate Azure-hosted Mistral specifically for summarization. Using a cheaper/faster model (e.g., Mistral Small, Haiku) could cut summarization costs by 5-10x.

2. **Retrieval parameters are hardcoded per plugin.** Values like `n_results`, score thresholds, and context limits need to be tuned in production based on observed retrieval quality — deploying code changes to adjust these is too slow for iterative tuning.

### Summarization LLM
- `SUMMARIZE_LLM_PROVIDER` — separate provider for ingestion summarization (default: falls back to `LLM_PROVIDER`)
- `SUMMARIZE_LLM_MODEL` — separate model (default: falls back to `LLM_MODEL`)
- `SUMMARIZE_LLM_API_KEY` — separate API key (default: falls back to `LLM_API_KEY`)
- `SUMMARIZE_LLM_TEMPERATURE` — typically lower for summarization (e.g., 0.1-0.3)

Implementation: create a second `LLMPort` instance in the plugin `__init__` when `SUMMARIZE_LLM_*` vars are set, pass it to `DocumentSummaryStep` and `BodyOfKnowledgeSummaryStep` instead of the main LLM.

### Retrieval parameters
- `EXPERT_N_RESULTS` — number of chunks retrieved by expert plugin (default: 5, currently hardcoded at 10)
- `GUIDANCE_N_RESULTS` — number of chunks per collection in guidance plugin (default: 5, currently hardcoded at 5)
- `MIN_RETRIEVAL_SCORE` — minimum `1.0 - distance` threshold (default: 0.0 = no filtering, see #14)
- `MAX_CONTEXT_CHARS` — total context budget before LLM call (default: 20000, see #15)

### Ingestion tuning
- `SUMMARY_CHUNK_THRESHOLD` — minimum chunks per document to trigger summarization (default: 3, currently hardcoded at line 134 of steps.py)

## What NOT to configure (tune in code, leave as constants)

These are implementation details that should be set through testing and baked into the code. Exposing them as configuration adds operational complexity without proportional benefit — they're unlikely to change between environments (dev/staging/prod) and misconfiguring them silently degrades quality.

| Parameter | Current Value | Where | Why not configurable |
|---|---|---|---|
| **Progressive budget ratio** (0.4 + 0.6 * progress) | 0.4 base, 0.6 range | `steps.py:46` | This is a summarization algorithm detail. The ratio determines how summary length grows as more chunks are processed. Changing it requires understanding the refine-pattern internals — exposing it invites misconfiguration with no clear production need. Tune once through testing. |
| **RecursiveCharacterTextSplitter separators** | LangChain defaults: `["\n\n", "\n", " ", ""]` | `steps.py:83-86` | The default separator hierarchy (paragraph → line → word → char) is well-tested and works for all content types in the pipeline. Custom separators are only needed for exotic formats (code, markdown tables) which aren't the primary content. |
| **EmbedStep/StoreStep batch_size** | 50 | `steps.py:250,287` | This is a throughput optimization, not a quality knob. The value depends on the embedding API's rate limits and ChromaDB's upsert performance — both are infrastructure concerns, not content concerns. If rate limiting becomes an issue, fix it in the adapter layer with backoff, not via configuration. |
| **Guidance DEFAULT_COLLECTIONS list** | 3 hardcoded collections | `guidance/plugin.py:18-22` | These are platform-specific knowledge bases (alkem.io, welcome.alkem.io, www.alkemio.org). They change when the platform's public knowledge structure changes, which is an architectural decision, not a tuning knob. |
| **Summary refine prompt templates** | DOCUMENT_REFINE_SYSTEM, etc. | `prompts.py:3-81` | Prompt engineering is a code-level concern. Prompts need testing, version control, and review — making them configurable via env vars bypasses all three. If prompts need to change per deployment, use a prompt management system, not env vars. |
| **Dedup logic in guidance** (top 5 unique sources) | 5 | `guidance/plugin.py:106` | This is tightly coupled with `GUIDANCE_N_RESULTS` and the number of collections. Exposing it separately creates a confusing 2-dimensional tuning space. Keep it as `min(n_results, 5)` or derive it from `GUIDANCE_N_RESULTS`. |
| **Score calculation formula** (`1.0 - distance`) | Cosine distance | `guidance/plugin.py:73` | This depends on the ChromaDB distance metric, not on business requirements. If you change the distance metric, the formula changes in code. |

## Files to change
- `core/config.py` — add `summarize_llm_*` fields, retrieval parameter fields
- `core/provider_factory.py` — create summarization LLM instance when configured
- `plugins/ingest_space/plugin.py` — pass separate summarization LLM to summary steps
- `plugins/ingest_website/plugin.py` — same
- `plugins/expert/plugin.py` — read `EXPERT_N_RESULTS` from config
- `plugins/guidance/plugin.py` — read `GUIDANCE_N_RESULTS` from config
- `core/domain/pipeline/steps.py:134` — read `SUMMARY_CHUNK_THRESHOLD` from constructor param

## Acceptance criteria
- [ ] Summarization steps can use a separate, cheaper LLM when `SUMMARIZE_LLM_*` env vars are set
- [ ] Falls back to main LLM when summarization-specific vars are not set
- [ ] `EXPERT_N_RESULTS` and `GUIDANCE_N_RESULTS` are configurable via env vars
- [ ] `SUMMARY_CHUNK_THRESHOLD` is configurable via constructor parameter (not hardcoded)
- [ ] Existing behavior unchanged when new env vars are not set (backward compatible)
- [ ] `.env.example` updated with all new variables and sensible defaults

### PR DESCRIPTION
**Pull request by valentinyanakiev**
## Summary

- **Separate summarization LLM**: Configure a cheaper model (`SUMMARIZE_LLM_PROVIDER`/`MODEL`/`API_KEY`) for document and body-of-knowledge summarization during ingestion, reducing costs 5-10x while keeping the main LLM for user-facing responses. Falls back gracefully when unconfigured.
- **Per-plugin retrieval parameters**: Externalize `EXPERT_N_RESULTS`, `EXPERT_MIN_SCORE`, `GUIDANCE_N_RESULTS`, `GUIDANCE_MIN_SCORE`, and `MAX_CONTEXT_CHARS` as env vars for production tuning without code changes. Context budget enforcement drops lowest-scoring chunks when total chars exceed the budget.
- **Configurable chunk threshold**: `SUMMARY_CHUNK_THRESHOLD` controls the minimum chunk count before document summarization triggers. Default `4` with `>=` preserves existing `> 3` behavior (FR-009).

All changes are additive — no port/adapter interface changes. Full backward compatibility when no new env vars are set.

## Files Changed

| File | Change |
|------|--------|
| `core/config.py` | 11 new config fields + validation rules + partial config warning |
| `main.py` | Startup config logging, summarization LLM wiring, per-plugin retrieval injection |
| `core/domain/pipeline/steps.py` | Configurable chunk threshold, model name logging |
| `core/adapters/langchain_llm.py` | Token-usage DEBUG logging |
| `plugins/expert/plugin.py` | `max_context_chars` enforcement |
| `plugins/guidance/plugin.py` | `n_results` param, `max_context_chars` enforcement |
| `plugins/ingest_website/plugin.py` | `summarize_llm` + `chunk_threshold` params |
| `plugins/ingest_space/plugin.py` | `summarize_llm` + `chunk_threshold` params |
| `.env.example` | All 11 new env vars documented |

## Test plan

- [x] 38 new tests across 4 test files (config validation, context budget, chunk threshold)
- [x] 256 total tests pass
- [x] Ruff lint: all checks passed
- [x] Pyright: 0 errors
- [x] Backward compatibility verified: all existing behavior unchanged with no new env vars set

Closes alkem-io/virtual-contributor#21

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  - Optional separate summarization LLM with automatic fallback to the main LLM.
  - Tunable per-plugin retrieval parameters (result counts, min scores), global context-character budget, and configurable summarization chunk threshold.
  - Context-budget enforcement that truncates oversized retrievals.

* **Chores**
  - Stronger configuration validation and startup logging with API-key masking; per-call model-name and token-usage logging added.

* **Documentation**
  - Spec, plan, quickstart, checklist, data-model, research, and tasks added.

* **Tests**
  - New tests for config validation, context budget, and summarization threshold.

### CODE REVIEW

File: core/domain/pipeline/steps.py [hunk 1]
```diff
@@ -106,17 +106,22 @@ async def execute(self, context: PipelineContext) -> None:
...
         self,
         llm_port: LLMPort,
         summary_length: int = 10000,
         concurrency: int = 8,
+        chunk_threshold: int = 4,
     ) -> None:
         self._llm = llm_port
         self._summary_length = summary_length
         self._concurrency = concurrency
+        self._chunk_threshold = chunk_threshold
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply core/domain/pipeline/steps.py:256]: Fixed in 6117506. Added `if chunk_threshold < 1: raise ValueError` guard at construction time.

File: specs/007-configurable-summarization/data-model.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+| `n_results` | `int` | `5` | Number of chunks to retrieve per collection |
+
+### Changed Behavior
+
+- Query calls use `self._n_results` instead of hardcoded `5`
+- Dedup limit uses `self._n_results` instead of hardcoded `5`
+
+## Entity: ExpertPlugin (unchanged)
+
+Already accepts `n_results` and `score_threshold` in constructor. No changes needed.
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply specs/007-configurable-summarization/data-model.md]: Fixed in 6117506. Updated the heading to "ExpertPlugin (modified)" and documented the new `max_context_chars` parameter.

File: specs/007-configurable-summarization/research.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+**Decision**: `MAX_CONTEXT_CHARS` applies per-plugin to the active plugin's merged retrieval results. Default: 20,000.
+**Rationale**: Consistent with microkernel isolation. The guidance plugin's 3-collection merge is the real "merged pool" the spec refers to.
+**Alternatives considered**: (a) Shared retrieval layer combining expert + guidance results — rejected (violates plugin isolation, requires architectural change far beyond scope).
+
+---
+
+### R5: SUMMARY_CHUNK_THRESHOLD — current behavior and configurability
+
+**Context**: FR-007 requires `SUMMARY_CHUNK_THRESHOLD` (default: 3). Currently `DocumentSummaryStep` hardcodes `if len(doc_chunks) > 3` (steps.py:134).
+
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply specs/007-configurable-summarization/research.md]: Fixed in 6117506. Updated the context line to say default 4, matching the resolution.

File: specs/007-configurable-summarization/spec.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+- **FR-004**: System MUST allow the number of retrieved chunks per collection for the guidance plugin to be configured via `GUIDANCE_N_RESULTS` (default: 5).
+- **FR-005**: System MUST allow per-plugin minimum retrieval score thresholds via `EXPERT_MIN_SCORE` (default: 0.3) and `GUIDANCE_MIN_SCORE` (default: 0.3). These defaults match the current `retrieval_score_threshold=0.3` behavior (FR-009). A score of 0.0 means no filtering.
+- **FR-006**: System MUST allow a maximum context character budget to be configured via `MAX_CONTEXT_CHARS` (default: 20,000). When the budget is exceeded, the system MUST merge all retrieved chunks from both expert and guidance plugins into a single pool and drop the lowest-scoring chunks globally until the total context fits within the budget.
+- **FR-007**: System MUST allow the minimum chunk count for triggering summarization to be configured via `SUMMARY_CHUNK_THRESHOLD` (default: 4). Documents with chunk count `>= SUMMARY_CHUNK_THRESHOLD` are summarized. The default of 4 preserves the current `> 3` behavior exactly (FR-009).
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply specs/007-configurable-summarization/spec.md]: Fixed in 6117506. Reworded FR-006 to clarify per-plugin budget enforcement (plugins are isolated containers). Also updated the matching clarification Q/A entry.

File: specs/007-configurable-summarization/spec.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+## Clarifications
+
+### Session 2026-04-06
+
+- Q: EXPERT_N_RESULTS default is 5 (FR-003) or 10 (User Story 2 scenario 5)? → A: Default is 5 (FR-003 is authoritative; User Story 2 scenario 5 corrected).
+- Q: When MAX_CONTEXT_CHARS is exceeded, how should the system truncate? → A: Drop lowest-scoring chunks first until under budget.
+- Q: Which providers must the summarization LLM support? → A: All providers the main LLM already supports (OpenAI, Mistral, Anthropic).
+- Q: Should the system log which model was used for each summarization call? → A: Yes, log model name + token count per call at INFO level.
+- Q: When the summarization LLM call fails, what should the system do? → A: Retry up to 3 times, then skip summarization and continue ingestion without a summary for that document.
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply specs/007-configurable-summarization/spec.md]: Fixed in 6117506. Updated the clarification to specify model name + document ID at INFO level and token usage at DEBUG level per FR-011.

**Comment by valentinyanakiev** [comment]
## Responses to CodeRabbit Nitpick Comments

### ✅ Fixed

- **`tests/test_config_retrieval.py` — Add invalid-value tests for `max_context_chars`**: Added tests for `max_context_chars=0` and negative values in 6117506.
- **`plugins/expert/plugin.py` — `dropped_chars` uses formatted vs raw lengths**: Fixed in 6117506. Now uses raw content lengths consistently for both budget enforcement and logging.

### Won't Fix (with rationale)

- **`plugins/guidance/plugin.py:75-82` — Remove redundant `n_results` local**: The local `n_results` variable is intentional. The `_query_collection` closure is passed to `asyncio.gather` — capturing a simple int via local variable is a cleaner closure pattern than capturing `self` (which the original code already used for `score_threshold` vs the query). Keeping it consistent with the existing code style.

- **`.env.example:44-63` — Reorder for dotenv-linter**: The current grouping is semantic (summarization LLM → retrieval params → budget → threshold), which is more readable for operators than alphabetical ordering. We don't use dotenv-linter in this project.

- **`tests/test_context_budget.py:5-30` — Prefer integration tests over helper**: The standalone `_apply_context_budget` tests validate the algorithm in isolation, which is their purpose. The file already includes plugin integration tests (`TestContextBudgetInExpertPlugin`, `TestContextBudgetInGuidancePlugin`) that verify plugin construction. Full `handle()` integration tests would require extensive mocking of async query pipelines and are better suited for end-to-end testing.
